In [ ]:
import pandas as pd
import numpy as np


In [ ]:
from google.colab import files
upload = files.upload()

Saving WA_Fn-UseC_-Telco-Customer-Churn.csv to WA_Fn-UseC_-Telco-Customer-Churn (1).csv


In [ ]:
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

Let's start Data Inspection


In [ ]:
df.shape

(7043, 21)

In [ ]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [ ]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [ ]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

From this data Inspection we analyzed some problems. so Let's solve the problems in Data Cleaning


In [ ]:
# checking the duplicate rows
df.duplicated().sum()

np.int64(0)

In [ ]:
#in total charges column there is like empty strings  " "

df["TotalCharges"].unique()[:10]

array(['29.85', '1889.5', '108.15', '1840.75', '151.65', '820.5',
       '1949.4', '301.9', '3046.05', '3487.95'], dtype=object)

In [ ]:
# Now wwe will turn to numeric column safely

df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors = "coerce")

In [ ]:
df["TotalCharges"].isna().sum()

np.int64(11)

In [ ]:
df = df.dropna(subset = ["TotalCharges"])

In [ ]:
df.shape

(7032, 21)

In [ ]:
#now we recheck data types once
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

In [ ]:
service_cols = ["MultipleLines", "OnlineSecurity", "DeviceProtection",
                "TechSupport", "StreamingTV", "StreamingMovies"]

for col in service_cols:
  print(f"{col}:{df[col].unique()}")
  print("-" * 40)

MultipleLines:['No phone service' 'No' 'Yes']
----------------------------------------
OnlineSecurity:['No' 'Yes' 'No internet service']
----------------------------------------
DeviceProtection:['No' 'Yes' 'No internet service']
----------------------------------------
TechSupport:['No' 'Yes' 'No internet service']
----------------------------------------
StreamingTV:['No' 'Yes' 'No internet service']
----------------------------------------
StreamingMovies:['No' 'Yes' 'No internet service']
----------------------------------------


In [ ]:
df[service_cols] = df[service_cols].replace({"No phone service": "No",
                                     "No internet service": "No"})

In [ ]:
for col in service_cols:
    print(f"{col}: {df[col].unique()}")
    print("-" * 40)


MultipleLines: ['No' 'Yes']
----------------------------------------
OnlineSecurity: ['No' 'Yes']
----------------------------------------
DeviceProtection: ['No' 'Yes']
----------------------------------------
TechSupport: ['No' 'Yes']
----------------------------------------
StreamingTV: ['No' 'Yes']
----------------------------------------
StreamingMovies: ['No' 'Yes']
----------------------------------------


In [ ]:
#Now we need to check senior citizen column for logical cleaning if needed

df["SeniorCitizen"].value_counts()

SeniorCitizen
0    5890
1    1142
Name: count, dtype: int64

In [ ]:
#Now data Celaning is completed. But we need to check every thing is cleaned or not.

df.isna().sum() #Checking na values



customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [ ]:
df.shape #Cheking shape of dataset


(7032, 21)

In [ ]:
df.dtypes #Checking datatypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

In [ ]:
for col in service_cols:
    print(col, df[col].unique())


MultipleLines ['No' 'Yes']
OnlineSecurity ['No' 'Yes']
DeviceProtection ['No' 'Yes']
TechSupport ['No' 'Yes']
StreamingTV ['No' 'Yes']
StreamingMovies ['No' 'Yes']


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

Now let's go to Exploratory data Analysis(EDA)

In [ ]:
df["Churn"].value_counts(normalize = True)

Churn
No     0.734215
Yes    0.265785
Name: proportion, dtype: float64

26.6% are Churned


In [ ]:
df["Churn"].value_counts()

Churn
No     5163
Yes    1869
Name: count, dtype: int64

In [ ]:
#we will check does contract type is effected churn

df.groupby("Contract")["Churn"].value_counts(normalize = True)

Contract        Churn
Month-to-month  No       0.572903
                Yes      0.427097
One year        No       0.887228
                Yes      0.112772
Two year        No       0.971513
                Yes      0.028487
Name: proportion, dtype: float64

In [ ]:
# Now we will analyze Numerical like tenure and Churn

df.groupby("Churn")["tenure"].mean()


Churn
No     37.650010
Yes    17.979133
Name: tenure, dtype: float64

“EDA shows that customers who churn have significantly lower average monthly charges compared to retained customers. This suggests that low-cost, short-term customers are more likely to leave the service.”

In [ ]:
df.groupby("Churn")["MonthlyCharges"].mean()


Churn
No     61.307408
Yes    74.441332
Name: MonthlyCharges, dtype: float64

Preparing data in a form that a machine learning model can understand by Feature Engineering.

In [ ]:
y = df["Churn"] # It is a target variable what we need to predict in model

In [ ]:
x =df.drop("Churn", axis = 1) #Removing target columns from inputs

If we didn't drop this column (or) Churn will stays in X Model Cheats and gives fake Accuracy.

In [ ]:
x = x.drop("customerID", axis = 1) # Removing ID Column

Removing the Customer ID column because it doesn't tells why customer churns. It just a unique label not behaviour

In [ ]:
#Now we are converting target variable into numbers

y = y.map({"Yes": 1, "No": 0}) # Converting to Binary form

Why Because ML Cannot learn from text. Binary classification needs binary values

In [ ]:
#Now we will identify which columns are in text
categorical_cols = x.select_dtypes(include = ["object"]).columns

In [ ]:
categorical_cols

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')

In [ ]:
#Now we will  do One - hot Encoding.To convert  catogorical to numerical columns
x =pd.get_dummies(x, columns = categorical_cols, drop_first = True)

#pd.get_dummies() -> converts text into binary (0/1)
#Now all columns in 'x' are numeric

Because to avoid redudancy, Multicollinearity and Improve model stability.

In [ ]:
#Now for better feature understanding we divide Tenure into buckets

x["tenure_group"] = pd.cut(
    x["tenure"],             #This Column we are grouping
    bins = [0, 12, 48, 72],  #Bucket boundaries
    labels = ["New", "Mid", "Long"]  #Human readable group names
)

This ranges are business friendly and commonly used.

In [ ]:
# Now we Find the median of MonthlyCharges
median_charge = x["MonthlyCharges"].median()

# Creating a binary feature:
# 1 = High charges
# 0 = Low charges
x["High_Monthly_Charges"] = (x["MonthlyCharges"] > median_charge).astype(int)


In [ ]:
# Encode the newly created tenure_group column
X = pd.get_dummies(x, columns=["tenure_group"], drop_first=True)


Now in this Feature engineering we added more features which i svery useful for ML model.

Now we will move to Train -Test Split

In [ ]:
# function to split data into train & test

from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X,                      # input features
    y,                      # target variable
    test_size=0.2,          # 20% data for testing
    random_state=42,        # ensures reproducibility
    stratify=y              # preserves churn ratio in train & test
)


In [ ]:
# We using logistic regression model for binary classification

from sklearn.linear_model import LogisticRegression


In [ ]:
log_model = LogisticRegression(max_iter=1000)


Why we created max_iter is if iterations ae too low model get fails and gives warnings.

In [ ]:
 # train the model using training data
log_model.fit(X_train, y_train)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

Here .fit() tells model to learn x_train for Input features Y_train for correct answers

In [ ]:
y_pred = log_model.predict(X_test) # It is predicting churn for unseen data

The main point is we Never predict on training data for evaluation.we Always predict on test data.

Now we successfully bulid the model

Now we are doing main step that is Evaluating the model. To check how the model's predictions are good by comparing actual values and predicted values

In [ ]:
# tool to compare actual vs predicted values
from sklearn.metrics import confusion_matrix


In [ ]:
# compare true vs predicted churn values
cm = confusion_matrix(y_test, y_pred)

In [ ]:
cm

array([[923, 110],
       [172, 202]])

In this particular output array
* 923 means it is True negitive and identified 923 customers are NOT churn
* 110 means False positive, model saying customers will churn, but they did not yet.
* 172 means False negitive, mdel missed 172 churners. This is most Important Number in Churn problems.
* 202 are True Positives, Model correctly Identifies 202 customers who actually Churned


923 + 110 + 172 + 202 = 1407
 It is a Quick sanity test, No Mistakes in Evaluation.

In [ ]:
#Now we will import accuracy function
from sklearn.metrics import accuracy_score  # metric to calculate overall correctness


In [ ]:
accuracy = accuracy_score(y_test, y_pred)  #calculate overall accuracy


In [ ]:
accuracy

0.7995735607675906

Majority of the customers do not churn, also predics non-churn well

* Although the model achieves around 80% accuracy, accuracy alone is insufficient for churn prediction because it hides the cost of missed churners

In [ ]:
# gives precision, recall, f1-score together

from sklearn.metrics import classification_report

In [ ]:
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.84      0.89      0.87      1033
           1       0.65      0.54      0.59       374

    accuracy                           0.80      1407
   macro avg       0.75      0.72      0.73      1407
weighted avg       0.79      0.80      0.79      1407



Class 0 → Not Churn
High precision & recall
Model is very good at identifying loyal customers
This explains the good accuracy
Class 1 → Churn (MOST IMPORTANT)
Recall ≈ 0.54
Model catches ~54% of churners
~46% churners are missed
Precision ≈ 0.65
When model predicts churn, it’s correct 65% of the time
This aligns perfectly with:

FN = 172 (missed churners)

TP = 202 (caught churners)

Now we Improve the model
because:
From your confusion matrix:

Missed churners (FN) = 172

Recall for churn ≈ 0.54

In [ ]:
#So we are telling the model pay more attention to Churn customers

log_model_bal = LogisticRegression(
    max_iter=1000,           # allow enough iterations to converge
    class_weight="balanced"  # give more importance to churn class
)


In [ ]:
# train the balanced model on training data
log_model_bal.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(class_weight='balanced', max_iter=1000)

In [ ]:
# predict churn using balanced model
y_pred_bal = log_model_bal.predict(X_test)

In [ ]:
y_pred_bal

array([0, 1, 0, ..., 0, 0, 0])

In [ ]:
# confusion matrix for balanced model
cm_bal = confusion_matrix(y_test, y_pred_bal)

In [ ]:
cm_bal

array([[729, 304],
       [ 78, 296]])

True Negatives (TN) = 729

Correctly predicted non-churn customers
Decreased compared to before (expected)

False Negatives (FN) = 78

Missed churners
THIS IS THE MOST IMPORTANT CHANGE
Before:
FN = 172
Now:
FN = 78

94 fewer churners missed

THis is the main point in churn prediction

In this "recall for churn" is really improved. Original model missed 46% churners. Balanced model misses ~21% churners.
THis is exactly what we want

But false positives increases. more loyal customers flagged as churn. I think this is acceptable trade off in churn problems.

In [ ]:
 # probability of churn (class 1)
y_prob_bal = log_model_bal.predict_proba(X_test)[:, 1]

In [ ]:
y_prob_bal

array([0.05687592, 0.80428277, 0.02103119, ..., 0.27837211, 0.05145609,
       0.02054191])

“If I randomly pick one churner and one non-churner, how often will the model rank the churner higher?”

In [ ]:
from sklearn.metrics import roc_auc_score  # metric to measure class-separation ability


In [ ]:
# calculate ROC–AUC for balanced model
roc_auc = roc_auc_score(y_test, y_prob_bal)

In [ ]:
roc_auc


np.float64(0.8340783554467285)

If we randomly pick one churner and one non-churner,
your model will correctly rank the churner 83.4% of the time.

That is very good for a churn prediction problem.

In [ ]:
#now we are choosing a new threshold lower than 0.5
threshold = 0.4


In [ ]:
y_pred_thresh = (y_prob_bal >= threshold).astype(int)


In [ ]:
y_pred_thresh

array([0, 1, 0, ..., 0, 0, 0])

In [ ]:
cm_thresh = confusion_matrix(y_test, y_pred_thresh)  # confusion matrix at threshold 0.4


In [ ]:
cm_thresh

array([[632, 401],
       [ 44, 330]])

“By tuning the decision threshold to 0.4, we significantly reduced missed churners from 172 to 44, achieving very high recall at the expense of increased false positives. This trade-off is acceptable in churn scenarios where retaining at-risk customers is more valuable than the cost of additional retention offers.”

True Negatives (TN) = 632

Correctly identified non-churn customers
Decreased further (expected)
False Positives (FP) = 401

Loyal customers predicted as churn
Increased

That means we need to do:

More customers will receive unnecessary retention offers

Cost increases, but customers stay

In [ ]:
#Now we import a Tree based model
from sklearn.tree import DecisionTreeClassifier  # decision tree model for classification


In [ ]:
dt_model = DecisionTreeClassifier(
    max_depth=5,              # limit tree depth to prevent overfitting
    class_weight="balanced",  # give more importance to churn class
    random_state=42           # reproducibility
)


In [ ]:
dt_model.fit(X_train, y_train)

DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=42)

In [ ]:
y_pred_dt = dt_model.predict(X_test)

In [ ]:
cm_dt = confusion_matrix(y_test, y_pred_dt)

In [ ]:
cm_dt

array([[700, 333],
       [ 81, 293]])

Decision Tree — number-by-number meaning
True Negatives (TN) = 700
Correctly predicted non-churn customers
Better than threshold-tuned logistic, worse than baseline logistic.

False Positives (FP) = 333
Predicted churn, but customer did not churn
False Negatives (FN) = 81

❌Missed churners

Compare:

Original Logistic → 172
Balanced Logistic → 78
Threshold 0.4 → 44
Decision Tree → 81
 Decision Tree is much better than baseline,
but worse than threshold-tuned logistic.

“The Decision Tree model improved churn detection compared to baseline logistic regression by capturing non-linear patterns, but the threshold-tuned logistic regression still achieved the best recall, making it more suitable for churn-focused business objectives.”

In [ ]:
from sklearn.ensemble import RandomForestClassifier


In [ ]:
rf_model = RandomForestClassifier(
    n_estimators=100,        # number of decision trees
    max_depth=8,             # limit depth to prevent overfitting
    class_weight="balanced", # handle churn class imbalance
    random_state=42          # reproducibility
)


In [ ]:
rf_model.fit(X_train, y_train)  # train the random forest on training data


RandomForestClassifier(class_weight='balanced', max_depth=8, random_state=42)

In [ ]:
y_pred_rf = rf_model.predict(X_test)

In [ ]:
cm_rf = confusion_matrix(y_test, y_pred_rf)

In [ ]:
cm_rf


array([[761, 272],
       [ 85, 289]])

True Negatives (TN) = 761

Correctly predicted non-churn customers
Better than Decision Tree and threshold-tuned logistic.
False Positives (FP) = 272

Loyal customers predicted as churn

Business meaning:

Fewer unnecessary offers than:

Balanced Logistic

Threshold-tuned Logistic

Better cost control

False Negatives (FN) = 85

Missed churners

Compare:

Original Logistic : 172

Balanced Logistic : 78

Threshold 0.4 : 44

Decision Tree : 81

Random Forest : 85

 Random Forest is much better than baseline,
but slightly worse than the best recall model (threshold-tuned logistic).